platform/ui/src/components/downloadDialog/DownloadDialog.js

```diff
@@ -0,0 +1,279 @@
+import React, { useEffect, useState } from 'react';
+import Modal from 'react-bootstrap-modal';
+
+import './DownloadDialog.styl';
+import { TextInput, Select } from '@ohif/ui';
+import { withTranslation } from '../../utils/LanguageProvider';
+
+const FILE_TYPE_OPTIONS = [
+  {
+    key: 'jpg',
+    value: 'jpg'
+  },
+  {
+    key: 'png',
+    value: 'png'
+  }
+];
+
+const DownloadDialog = ({
+  t,
+  isOpen,
+  activeViewport,
+  onClose,
+  updateViewportPreview,
+  enableViewport,
+  disableViewport,
+  toggleAnnotations,
+  loadImage,
+  downloadBlob,
+  defaultSize,
+  minimumSize,
+  canvasClass
+}) => {
+  const [filename, setFilename] = useState('image');
+  const [fileType, setFileType] = useState('jpg');
+
+  const [height, setHeight] = useState(defaultSize);
+  const [width, setWidth] = useState(defaultSize);
+
+  const [showAnnotations, setShowAnnotations] = useState(true);
+
+  const [keepAspect, setKeepAspect] = useState(true);
+  const [lastImage, setLastImage] = useState();
+
+  const [viewportElement, setViewportElement] = useState();
+  const [viewportElementHeight, setViewportElementHeight] = useState(minimumSize);
+  const [viewportElementWidth, setViewportElementWidth] = useState(minimumSize);
+
+  const [downloadCanvas, setDownloadCanvas] = useState();
+  const [downloadCanvasHeight, setDownloadCanvasHeight] = useState(minimumSize);
+  const [downloadCanvasWidth, setDownloadCanvasWidth] = useState(minimumSize);
+
+  const [viewportPreview, setViewportPreview] = useState();
+  const [viewportPreviewSrc, setViewportPreviewSrc] = useState();
+  const [viewportPreviewHeight, setViewportPreviewHeight] = useState(minimumSize);
+  const [viewportPreviewWidth, setViewportPreviewWidth] = useState(minimumSize);
+
+  useEffect(() => {
+    enableViewport(viewportElement);
+
+    return () => {
+      disableViewport(viewportElement);
+
+      setHeight(defaultSize);
+      setWidth(defaultSize);
+    };
+  }, [viewportElement]);
+
+  const loadAndUpdateViewports = async () => {
+    const {
+      image,
+      width: scaledWidth,
+      height: scaledHeight
+    } = await loadImage(activeViewport, viewportElement, width, height);
+
+    setLastImage(image);
+
+    toggleAnnotations(showAnnotations, viewportElement);
+
+    setViewportElementHeight(scaledHeight);
+    setViewportElementWidth(scaledWidth);
+    setDownloadCanvasHeight(scaledHeight);
+    setDownloadCanvasWidth(scaledWidth);
+
+    const {
+      dataUrl,
+      width: viewportElementWidth,
+      height: viewportElementHeight
+    } = await updateViewportPreview(viewportElement, downloadCanvas, fileType);
+
+    setViewportPreviewSrc(dataUrl);
+    setViewportPreviewHeight(viewportElementHeight);
+    setViewportPreviewWidth(viewportElementWidth);
```

Comment:
If you're always using these three together, you can do:

```
const [viewportPreview, setViewportPreview] = useState({
  src, // undef or null
  height,
  width
});

...

setViewportPreview({
  src: dataUrl,
  height: viewportElementHeight,
  width: viewportElementWidth,
});

Reply:
Yeah, a lot better! setting big objects sometimes scares me a little but in this case its good. Done, thx.